### ISSUE TITLE
Better handling of wrong app URL when creating section

### ISSUE THREAD

**Issue by jamesscottbrown** [ove/ove#71]
Whilst running OVE on my local machine using docker-compose, I tried to create an image section and thoughtlessly gave an app URL with the hostname `localhost` rather than my IP address :

```bash
curl --header "Content-Type: application/json" --request POST --data '{"app": {"url": "http://localhost:8082","states": {"load": {"tileSources": "https://openseadragon.github.io/example-images/highsmith/highsmith.dzi"}}}, "space": "LocalFour", "h": 500, "w": 500, "y": 0, "x": 0}' http://localhost:8080/section
```

This failed, but in a way that didn't immediately reveal what was wrong: the API call to  `/section`  returned an `id` as if it had worked correctly, but the Web Console in the browser windows displaying the OVE client pages and the image apps's controller gave a vague warning ("Lost websocket connection attempting to reconnect").

Perhaps OVE core should test if the specified app URL is reachable when trying to create a section, and report an error if it is not.

**Comment by oserban** [comment]
This would not make a big difference in practice because the resource needs to be available from the viewer pov.

**Comment by senakafdo** [comment]
@jamesscottbrown, there is nothing wrong in the way it is being handled @ backend, because your choice of URL is valid and it has no way to figure this out.

The loss of websocket connection is not quite related to what you explain. That happens only when you are unable to reach OVE, which is provided by the OVE_HOST url. If what you explained has happened, you would instead see a dead page on the browser, since the browser would not be able to load http://localhost:8082/view.html within its iFrame. I believe that what probably caused your WebSocket connection to be lost is OVE itself crashing for some odd reason.

If you can reproduce the issue consistently, can you have a look at your PM2 logs and share the exceptions/errors that you see in it?

**Comment by senakafdo** [comment]
The bug is in OVE core - I reproduced it. There is no way to tell you that you got your URL wrong, and if OVE actually got a proper URL and the browser cannot reach it, as you expected, then you would not get this far, because the page will never load in the first place. What you are seeing here is a result of OVE crashing and restarting in the background. It happens so fast to an extent where you'd probably think it was a WebSocket connection issue. The PM2 logs has the proper error in it.

```sh
14|@ove/ov | events.js:183
14|@ove/ov |       throw er; // Unhandled 'error' event
14|@ove/ov |       ^
14|@ove/ov | Error: connect ETIMEDOUT 1.2.3.4:8094
14|@ove/ov |     at Object._errnoException (util.js:1022:11)
14|@ove/ov |     at _exceptionWithHostPort (util.js:1044:20)
14|@ove/ov |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1198:14)
14|@ove/ov | npm
14|@ove/ov |  ERR! code ELIFECYCLE
14|@ove/ov | npm
14|@ove/ov |  ERR! errno 1
14|@ove/ov | npm 
14|@ove/ov | ERR! @ove/ove-core@0.2.0 start: `node dist/index.js --color`
14|@ove/ov | npm ERR! Exit status 1
14|@ove/ov | npm ERR! 
14|@ove/ov | npm ERR! Failed at the @ove/ove-core@0.2.0 start script.
14|@ove/ov | npm ERR!
14|@ove/ov |  This is probably not a problem with npm. There is likely additional logging output above.
14|@ove/ov | npm ERR! A complete log of this run can be found in:
14|@ove/ov | npm ERR!     /Users/senaka/.npm/_logs/2019-01-18T22_37_11_589Z-debug.log
PM2        | App [@ove/ove-core] with id [14] and pid [42441], exited with code [1] via signal [SIGINT]
PM2        | Starting execution sequence in -fork mode- for app name:@ove/ove-core id:14
PM2        | App name:@ove/ove-core id:14 online
14|@ove/ov | > @ove/ove-core@0.2.0 start /Users/senaka/University/Work/DSI/ove/packages/ove-core
14|@ove/ov | > node dist/index.js --color
14|@ove/ov | [DEBUG] 18/01/2019, 10:37:16.016 pm - OVE              : Starting OVE Core
```

**Comment by senakafdo** [comment]
So, the problem here is that the request module runs our calls asynchronously. There will be no calls made to the URL that is provided when a section is created, unless you try to cache a state when creating a section. Even that should happen asynchronously and the request will return immediately, with a valid section-id. 

Now, when Node.js actually tries to make the request, it can fail due to one of many reasons: `ECONNRESET`, `ENOTFOUND`, `ESOCKETTIMEDOUT`, `ETIMEDOUT`, `ECONNREFUSED`, `EHOSTUNREACH` etc. These errors cannot be handled inside a try/catch block since node is asynchronous. Also, it is not a good idea to introduce a synchronous call during the creation of a section to test whether the URL is valid, as it will introduce a bottleneck. Also, it will not fix all issues such as a URL being temporarily unavailable due an application being restarted etc.

Therefore, the only and proper fix is to introduce some request-specific error handling. This is the basis of the fix in 07d3a08.
